### ISSUE TITLE
no-redeclare throws errors on overloaded functions

### ISSUE THREAD

**Issue by j-f1** [eslint/typescript-eslint-parser#402]
**What version of TypeScript are you using?**

2.5.3

**What version of `typescript-eslint-parser` are you using?**

8.0.0

**What code were you trying to parse?**

```ts
export async function readFile(
  filename: string,
  options?: { flag?: string }
): Promise<Buffer>
export async function readFile(
  filename: string,
  options?: { encoding: BufferEncoding; flag?: string }
): Promise<string>
export async function readFile(
  filename: string,
  options?: { encoding?: string; flag?: string }
): Promise<Buffer | string> {
  // ...
}
```

**What did you expect to happen?**

No errors — `readFile` is only declared once.

**What happened?**

```
'readFile' is already defined.
```

**Comment by JamesHenry** [comment]
We should hopefully be able to solve this with upcoming custom Referencer work

**Comment by JamesHenry** [comment]
This was finally fixed by https://github.com/eslint/typescript-eslint-parser/pull/412. I added the source above as an integration test.
